core-runner/src/ext.rs

```diff
@@ -53,14 +54,43 @@ impl Ext {
 }
 
 impl EnvExt for Ext {
-    fn alloc(&mut self, pages: PageNumber) -> Result<PageNumber, &'static str> {
-        self.gas(pages.raw() * self.alloc_cost as u32)?;
+    fn alloc(&mut self, pages_num: PageNumber) -> Result<PageNumber, &'static str> {
+        // Greedly sub gas for allocations
+        self.gas(pages_num.raw() * self.alloc_cost as u32)?;
+        // Greedly sub gas for grow
+        self.gas(pages_num.raw() * self.mem_grow_cost as u32)?;
+
+        let old_mem_size = self.memory_context.memory().size().raw();
 
         let result = self
             .memory_context
-            .alloc(pages)
+            .alloc(pages_num)
             .map_err(|_e| "Allocation error");
 
+        if result.is_err() {
+            return self.return_with_tracing(result);
+        }
+
+        // Returns back greedly used gas for grow
+        let new_mem_size = self.memory_context.memory().size().raw();
+        let grow_pages_num = new_mem_size - old_mem_size;
+        let mut gas_to_return_back = self.mem_grow_cost * (pages_num.raw() - grow_pages_num) as u64;
```

Comment:
I'd implement `Add` trait etc for `PageNumber` to avoid these multiple `raw()` and vice versa but we can do it as a refactoring later.

Reply:
`Add` and `Sub` traits for `PageNumber` have already been implemented.

Reply:
Yes, I suppose it's not good idea to implement it in this review. I may brake something=)